extras/rya.forwardchain/src/main/java/org/apache/rya/forwardchain/rule/SpinConstructRule.java

```diff
@@ -0,0 +1,344 @@
+        graphQuery.getTupleExpr().visit(new TypeRequirementVisitor("this", type));
+        // Find all statement patterns that could trigger this rule
+        AntecedentVisitor aVisitor = new AntecedentVisitor();
+        graphQuery.getTupleExpr().visit(aVisitor);
+        antecedentStatementPatterns = aVisitor.getAntecedents();
+        // Construct statement patterns for all possible conclusions of this rule
+        ConstructConsequentVisitor cVisitor = new ConstructConsequentVisitor();
+        graphQuery.getTupleExpr().visit(cVisitor);
+        consequentStatementPatterns = cVisitor.getConsequents();
+    }
+
+    /**
+     * Get the URI or bnode associated with this rule in the data.
+     * @return The rule's identifier.
+     */
+    public Resource getId() {
+        return ruleId;
+    }
+
+    @Override
+    public String toString() {
+        return "SpinConstructRule{" + ruleId.stringValue() + "}";
+    }
+
+    @Override
+    public ParsedGraphQuery getQuery() {
+        return graphQuery;
+    }
+
+    @Override
+    public boolean canConclude(StatementPattern sp) {
+        Preconditions.checkNotNull(sp);
+        Value s1 = getVarValue(sp.getSubjectVar());
+        Value p1 = getVarValue(sp.getPredicateVar());
+        Value o1 = getVarValue(sp.getObjectVar());
+        Value c1 = getVarValue(sp.getContextVar());
+        for (StatementPattern consequent : consequentStatementPatterns) {
+            Value s2 = getVarValue(consequent.getSubjectVar());
+            Value p2 = getVarValue(consequent.getPredicateVar());
+            Value o2 = getVarValue(consequent.getObjectVar());
+            Value c2 = getVarValue(consequent.getContextVar());
+            if ((s1 == null || s2 == null || s1.equals(s2))
+                    && (p1 == null || p2 == null || p1.equals(p2))
+                    && (o1 == null || o2 == null || o1.equals(o2))
+                    && (c1 == null || c2 == null || c1.equals(c2))) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public Collection<StatementPattern> getAntecedentPatterns() {
+        return antecedentStatementPatterns;
+    }
+
+    @Override
+    public Collection<StatementPattern> getConsequentPatterns() {
+        return consequentStatementPatterns;
+    }
+
+    @Override
+    public long execute(AbstractRuleExecutionStrategy strategy,
+            StatementMetadata metadata) throws ForwardChainException {
+        metadata.addMetadata(ForwardChainConstants.RYA_DERIVATION_RULE,
+                RdfToRyaConversions.convertResource(ruleId));
+        return super.execute(strategy, metadata);
+    }
+
+    private static Value getVarValue(Var var) {
+        return var == null ? null : var.getValue();
+    }
+
+    private static class TypeRequirementVisitor extends QueryModelVisitorBase<RuntimeException> {
+        private static final Var RDF_TYPE_VAR = new Var("-const-" + RDF.TYPE.stringValue(), RDF.TYPE);
```

Comment:
The hardcoded "-const-"'s will need to go away once RDF4J is added.  But that should be handled by [RYA-405](https://github.com/apache/incubator-rya/pull/245) so no worries here.